proposals/0000-package-dependency-import-syntax.md

```diff
@@ -0,0 +1,155 @@
+```swift
+let package = Package(
+    name: "swift-tools-support-core",
+    products: [
+        .library(
+            name: "SwiftToolsSupport",
+            type: .dynamic,
+            targets: ["TSCBasic", "TSCUtility"]),
+        .library(
+            name: "SwiftToolsSupport-auto",
+            targets: ["TSCBasic", "TSCUtility"]),
+
+        .library(
+            name: "TSCTestSupport",
+            targets: ["TSCTestSupport"]),
+        .executable(
+            name: "TSCTestSupportExecutable",
+            targets: ["TSCTestSupportExecutable"]),
+    ],
+    // Rest of Package.swift truncated for brevity .....
+}
+```
+
+With the above Package descriptions, the `SwiftToolsSupport` product does not match a target within the package description it must be specified explicitly.
+
+```swift
+@package(url: "https://github.com/apple/swift-tools-support-core.git", products: ["SwiftToolsSupport"]) 
+import SwiftToolsSupport 
+```
+
+#### Dependency version requirements
+
+The `@package(..)` annotation would also allow a developer to specify requirements to ensure your scripts are reproducible. The syntax will contain the same level of control and specificity as `PackageDependencyDescription.Requirement` which should feel like a natural to developers familiar with Swift Package Manager manifest syntax.  
+
+*Revision*
+```swift
+@package(url: "https://github.com/jpsim/Yams.git", .revision( "065675b3d1364a6f63b94a9c89be2e9ed0a4c3a1")) 
+import Yams 
+```
+
+*Branch*
+```swift
+@package(url: "https://github.com/jpsim/Yams.git", .branch("releases/1.0")) 
+import Yams 
+```
+
+*Exact Version*
+```swift
+@package(url: "https://github.com/jpsim/Yams.git", .exact("1.0")) 
+import Yams 
+```
+
+*Range of version*
+```swift
+@package(url: "https://github.com/jpsim/Yams.git", .from("1.0")) 
+import Yams 
+```
+
+*Local dependency*
+Absolute path
+```swift
+@package(path: "/Users/username/Yams") 
+import Yams 
+```
+
+Relative path
+```swift
+@package(path: "dependencies/Yams") 
+import Yams 
+```
+
+### Integration with Swift Compiler 
+
+The `@package()` syntax will need to be added to the AST definition and compiler front-end will require changes to parse this syntax.
+
+### Integration with SwiftPM
+
+#### Execution
+Scripts that utilize the `@package()` attribute can be executed by `swift run` and managed by Swift package manager.
+
+Swift package manager will utilize the changes to the compiler front-end to access packages referenced using the `@package` attribute.
+
+#### Package resolution 
+
+Once package dependencies are known, SwiftPM will use it's existing logic for resolving those dependencies and outputting a ".resolved" file to ensure future invocations are reproducible.
+
+The name of the resolved file will be `script_name.resolved`.
+
+#### Built products
+
+Products built from scripts will be located in common per-user location `~/.swiftpm/scripts/...`
```

Comment:
I'd like to hear some commentary about caching.

Will every execution of the script re-resolve? Thats fairly expensive, if so, and a shame. If not, how would users indicate they want to re-resolve?

In any case, resolutions which end up with the same package graph should definitely have their binary artifacts cached.

This mechanism doesn't afford any place to put a `Package.resolved` file, which to some degree makes it riskier from a security perspective. Not sure there is anything we can do about that, but a tad unfortunate...

Reply:
FWIW, I would probably propose that we be willing to cache resolution based on the timestamp of the file and perhaps other heuristic criteria, at the risk of being a little more magical. Basically, I think there is a massive amount of value in making sure that this operation is super fast when used as a shebang line, which means we will have to guarantee that the common path can end up fully cache.

Reply:
The latest version of the proposal mentions the existence of a local cache for dependencies that will also be able to store `Package.resolved` files. 

I agree 100% that speed of execution and avoiding re-resolving is a top consideration that needs to be taken during implementation